pubsub.go

```diff
@@ -660,14 +690,27 @@ func (p *PubSub) validate(vals []*topicVal, src peer.ID, msg *Message) {
 		}
 	}
 
+	// we can mark the message as seen now that we have verified the signature
+	// and avoid invoking user validators more than once
+	id := msgID(msg.Message)
+	if !p.markSeen(id) {
+		return
+	}
+
 	if len(vals) > 0 {
-		if !p.validateTopic(vals, src, msg) {
-			log.Warningf("message validation failed; dropping message from %s", src)
-			return
+		select {
+		case p.validateThrottle <- struct{}{}:
+			go func() {
+				p.doValidateTopic(vals, src, msg)
```

Comment:
Doesn't this defeat the point of having a worker pool? Is there any reason to spawn this as a goroutine?

Reply:
Yes, it kind of does. But at the moment all registered validators are asynchronous which means they can block, and we need to execute them in a goroutine.

Having said that, this is an artifact of WIP; I want to separate the validators to synchronous and asynchronous, and execute the synchronous ones inline.
The asynchronous can be in a goroutine or even better in an asynchronous api that doesn't use throttles at all.

Reply:
Actually, since we are going to spawn a goroutine for async validators, we can still use the throttles.

Reply:
Hm. But why not just have 1024 synchronous validators (well, dynamically adjusting)?

Reply:
The asynchronous validators can block at will, I don't think we should have pinned goroutines for them. The cost of actually spawning the goroutine shouldn't be too high.

Also, if we had that many synchronous goroutines we could overtax the system doing synchronous validation.
That is we could end up validating signatures for 1024 messages concurrently which is not the intention (oversubscribes the cpu by a lot).
Perhaps we could have a second pool with prespawned (or dynamically adjustable) goroutines for the async validators but I don't think we gain much for the complexity.

Reply:
I see. So we're assuming that some validators will be CPU bound and others will won't. Async validators are those non-CPU bound validators.

This is really where async/await shines.

>  Perhaps we could have a second pool with prespawned (or dynamically adjustable) goroutines for the async validators but I don't think we gain much for the complexity.

We can revisit this later.

Reply:
We might want to allow multiple validators per topic: one could register an inline validator that fails fast with basic structural validation and an async validator that fetches stuff from the network for deep validation.

Reply:
Ultimately I'd like to see something like a [claim-check pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/StoreInLibrary.html) here, but we can iterate towards that later. In summary:

1. The validator pipeline sends the message to the validator via a channel. The message includes the message ID, a deadline until which a result will be accepted, and a response channel.
2. The validator is free to manage its own goroutine pool, without the pipeline making assumptions whether the task is CPU or IO-bound.
3. The validator performs validation.
4. The validator returns the result to the pipeline via a channel, reporting the message ID and the result (OK/KO).